nixos/modules/services/matrix/mautrix-meta.nix

```diff
@@ -83,37 +78,43 @@ in {
                   address = "http://${config.settings.appservice.hostname}:${toString config.settings.appservice.port}";
                 };
 
-                meta = {
-                  mode = "";
-                };
-
                 bridge = {
-                  # Enable encryption by default to make the bridge more secure
-                  encryption = {
-                    allow = true;
-                    default = true;
-                    require = true;
+                  permissions = {};
+                };
 
-                    # Recommended options from mautrix documentation
-                    # for additional security.
-                    delete_keys = {
-                      dont_store_outbound = true;
-                      ratchet_on_decrypt = true;
-                      delete_fully_used_on_decrypt = true;
-                      delete_prev_on_new_session = true;
-                      delete_on_device_delete = true;
-                      periodically_delete_expired = true;
-                      delete_outdated_inbound = true;
-                    };
+                database = {
+                  type = "sqlite3-fk-wal";
+                  uri = "file:${fullDataDir config}/mautrix-meta.db?_txlock=immediate";
+                };
 
-                    verification_levels = {
-                      receive = "cross-signed-tofu";
-                      send = "cross-signed-tofu";
-                      share = "cross-signed-tofu";
-                    };
+                # Enable encryption by default to make the bridge more secure
+                encryption = {
+                  allow = true;
+                  default = true;
+                  require = true;
+
+                  # Recommended options from mautrix documentation
+                  # for additional security.
+                  delete_keys = {
+                    dont_store_outbound = true;
+                    ratchet_on_decrypt = true;
+                    delete_fully_used_on_decrypt = true;
+                    delete_prev_on_new_session = true;
+                    delete_on_device_delete = true;
+                    periodically_delete_expired = true;
+                    delete_outdated_inbound = true;
                   };
 
-                  permissions = {};
+                  # TODO: This effectively disables encryption. But this is the value provided when a <0.4 config is migrated. Changing it will corrupt the database.
+                  # https://github.com/mautrix/meta/blob/f5440b05aac125b4c95b1af85635a717cbc6dd0e/cmd/mautrix-meta/legacymigrate.go#L24
+                  # If you wish to encrypt the local database you should set this to an environment variable substitution and reset the bridge or somehow migrate the DB.
+                  pickle_key = "mautrix.bridge.e2ee";
```

Comment:
The config generator generates this randomly, but the migrator uses a fixed value. I think we could have generated this randomly if we shipped the module update at the same time as the package update but that ship has sailed.

However in addition to this there are two other values that are randomly generated each service start. They are `direct_media.server_key` and `public_media.signing_key`. I haven't yet noticed any effect of not preserving them but I haven't looked into it. I presume that we will want to preserve them in the same way that we currently preserve the `as_token`. (I really hate this state/secrets in config file crap...)

Reply:
What about handling this value similarly to what is done for other generated tokens - as_token and hs_token from the registration file? As in - read the config, see if the value is there. If not, do nothing. If yes, store it, and if it is not populated by envsubsting new config, put it to the resulted config.

This way the user won't have to do anything if it's generated. Also it solves the problem with migration, since user has this pickle key in the config, so it will just get copied as it was made from the migrator.

Btw. I have trouble finding answer to this - prior to 0.4 there was no real encryption of the database? If so, that seems to me like users coming from <0.4 are at least not losing anything if their db is still not encrypted.

Reply:
I don't think that is the best idea because this information is critical to recover the database. So if the user isn't backing up the config file then it would make backups of the rest of the application state broken. So I think it is best to have some sort of opt-in to ensure that they are aware of what they are doing.

In general I find this application-specific encryption kind of silly, as the key is sitting right next to the DB anyways. Really the only value it would have is if the key isn't stored in the same backups as the database itself. So generating a random key and expecting it to be included in the backup provides very little value over using a fixed key. It also removes the ability for the user to set up a separately stored key that is kept separate from the backup (unless we add extra complexity to support both modes).

So my thought is:

1. If you want to protect the DB you should be using disk encryption and backup encryption.
2. Assuming that you have the above encryption the application-level encryption is almost zero improvement.
3. If you want extra security then you can configure the key yourself and ensure that it doesn't end up in the backups.

So with that the fixed key by default seems pretty reasonable.